backoff.c

```diff
@@ -0,0 +1,87 @@
+#include "common.h"
+
+#include <ext/standard/php_rand.h>
+#include <ext/standard/php_mt_rand.h>
+
+#include "backoff.h"
+
+zend_long random_range(zend_long min, zend_long max) {
+    if (max <= min) {
+        max = min + 1;
+    }
+
+  return php_mt_rand_range(min, max);
+}
+
+zend_long default_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = retry_index ? self->base : random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long constant_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = self->base;
+    return MIN(self->cap, backoff);
+}
+
+zend_long uniform_backoff(struct Backoff *self, int retry_index) {
+  zend_long backoff = random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long exponential_backoff(struct Backoff *self, int retry_index) {
```

Comment:
Do we really need exponential backoff if it is always worse than jitter backoff? Is there any use cases why users may choose it?

Reply:
I'm honestly not sure.  Maybe we could reorder the array of options to put the best overall algorithms first?

Reply:
Or maybe we could remove useless algorithms? :smile:

Reply:
I think you make a good point.


@nbraun-amazon thoughts?

Reply:
I'd say leave the current option first, then reorder. I did it in this new revision.

Reply:
@yatsukhnenko I don't mind leaving them in.  I can see a use for testing since the non-jitter algorithms are deterministic.